### PR TITLE
Receiving new Challenges

### PR DESCRIPTION
**Pull request by ekengni**
Client will now receive a new unique challenge when one challenge is complete
- User now tracks completed challenges  
- ChallengeServlet will now update User's current, completed, and status of challenges, when a put request is sent.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ChallengesServlet.java [hunk 1]
```diff
@@ -92,16 +96,71 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.getWriter().println(json);
   }
 
+  /* remove challenge id from user status map
+     add new challenge id to user status map
+     add challenge id to user completed challenge map */
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (natalie-abreu) [inline src/main/java/com/google/sps/servlets/ChallengesServlet.java:104]: Is this replacing the functionality related to updating statuses in the user servlet? Or only when a completely new challenge is requested? If it overlaps with the user serlvet probably want to remove the part it is replacing

Comment (ekengni) [reply src/main/java/com/google/sps/servlets/ChallengesServlet.java:104]: oh no this is only when a new challenge is requested.

Comment (ekengni) [reply src/main/java/com/google/sps/servlets/ChallengesServlet.java:104]: i'll clarify my comment in the next commit

Comment (ekengni) [reply src/main/java/com/google/sps/servlets/ChallengesServlet.java:104]: So the doPut main purpose is to update User's 'completed challenges' and find and add a new challenge to 'challenge statuses'.

File: src/main/java/com/google/sps/servlets/ChallengesServlet.java [hunk 2]
```diff
@@ -92,16 +101,69 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
   }
 
+  //This function Update's User's Current, and Completed challenges
+  // along with statuses
+  private void updateUserChallenges(User user, String compl_id, String cur_id) {
+    HashMap<String, Integer> challenge_statuses = user.getChallengeStatuses();
+    HashSet<String> completed_challenges = user.getCompletedChallenges();
+    user.setCurrentChallenge(cur_id);
+
+    //challenge_statuses.remove(compl_id);
```
Comment (lazau-g) [inline src/main/java/com/google/sps/servlets/ChallengesServlet.java:149]: nit: remove

File: src/main/java/com/google/sps/servlets/ChallengesServlet.java [hunk 3]
```diff
@@ -92,16 +101,69 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    user.appendToCompletedChallenges(compl_id);
+    
+    //add new challenge to challenge status.
+    for(String challenge_id : ChallengeData.CHALLENGES_MAP.keySet()){
```
Comment (lazau-g) [inline src/main/java/com/google/sps/servlets/ChallengesServlet.java:153]: It seems like you're initializing user's challenge_status for all challenges in CHALLENGES_MAP. Wouldn't it make more sense to put that code into where the User is initialized? I.e. https://github.com/googleinterns/step72-2020/blob/e0fbc902167e97e4ba1699cb05c7b04d6414bdea/src/main/java/com/google/sps/servlets/UserServlet.java#L132 should read `.setChallengeStatuses(ChallengeData.DEFAULT_CHALLENGES_AND_STATUS)` where `ChallangeData.DEFAULT_CHALLENGES_AND_STATUS` should be similar to https://github.com/googleinterns/step72-2020/blob/e0fbc902167e97e4ba1699cb05c7b04d6414bdea/src/main/java/com/google/sps/servlets/UserServlet.java#L255 but modified to include all challenges present in `ChallangeData.CHALLENGES_MAP`.

**Pull request merged**